sharktank/sharktank/models/llm/testing.py

```diff
@@ -1,93 +1,185 @@
-    cache_state = [torch.rand_like(cache_state[0])]
-    return OrderedDict(
-        [
-            ("tokens", decode_token_ids),
-            ("seq_lens", seq_lens),
-            ("start_positions", start_positions),
-            ("seq_block_ids", seq_block_ids),
-            ("cache_state", cache_state),
-        ]
-    )
+    return token_ids_list
 
 
-def make_random_prefill_args(
-    model: PagedLlmModelV1, batch_size: int
-) -> OrderedDict[str, Any]:
-    seq_lens = torch.randint(
-        size=[batch_size],
-        low=1,
-        high=min(
-            2 * model.config.block_seq_stride,
-            model.config.hp.context_length,
+class LlmPerplexityCompare:
+    """Compare the perplexity of one implementation against another.
+    Checks that the 2 models' results do not deviate from each other."""
+
+    def __init__(
+        self,
+        make_target_model: Callable[[], LlmInstance],
+        make_reference_model: Callable[[], LlmInstance],
+        rtol: float,
+        atol: float,
+    ):
+        self.make_target_model = make_target_model
+        self.make_reference_model = make_reference_model
+        self.rtol = rtol
+        self.atol = atol
+
+    def run_and_assert_close(self, tokens: list[list[int]]):
+        reference_prefill_results, reference_decode_results = self._run(
+            self.make_reference_model, tokens=tokens
         )
-        - 1,
-        dtype=torch.int64,
-        device=model.device,
+        target_prefill_results, target_decode_results = self._run(
+            self.make_target_model, tokens=tokens
+        )
+
+        self._assert_close(target_prefill_results, reference_prefill_results)
+        self._assert_close(target_decode_results, reference_decode_results)
+
+    def _run(
+        self,
+        make_model: Callable[[], LlmInstance],
+        tokens: list[list[int]],
+    ):
+        instance = make_model()
+        perplexity_eval = instance.make_perplexity_eval()
+        prefill_results = perplexity_eval.prefill_cross_entropy(tokens)
+        decode_results = perplexity_eval.decode_cross_entropy(tokens)
+        assert all(result.valid for result in prefill_results)
+        assert all(result.valid for result in decode_results)
+        return prefill_results, decode_results
+
+    def _assert_close(
+        self,
+        actual: list[LlmPerplexityEval.Result],
+        expected: list[LlmPerplexityEval.Result],
+    ):
+        actual_scores = torch.tensor([r.score for r in actual], dtype=torch.float32)
+        expected_scores = torch.tensor([r.score for r in expected], dtype=torch.float32)
+        assert_tensor_close(
+            actual_scores, expected_scores, rtol=self.rtol, atol=self.atol
+        )
+
+
+def run_perplexity_test_pipeline_parallel_eager_vs_eager(
+    reference_theta: Theta,
+    reference_config: LlamaModelConfig,
+    tokens: list[list[int]],
+    atol: float = 0,
+    rtol: float = 0,
```

Comment:
I know I originally asked this to be paramaterized, but I was wrong.

These should be hardcoded as 0 when calling `LlmPerplexityCompare()` below.

Reply:
I hardcoded them in `run_perplexity_test_pipeline_parallel_eager_vs_eager`.